ADAL/src/ui/mac/ADAuthenticationViewController.m

```diff
@@ -0,0 +1,212 @@
+        windowRect = [[NSScreen mainScreen] frame];
+    }
+    
+    // Calculate the center of the current main window so we can position our window in the center of it
+    NSRect centerRect = _CenterRect(windowRect, NSMakeRect(0, 0, DEFAULT_WINDOW_WIDTH, DEFAULT_WINDOW_HEIGHT));
+    
+    NSWindow* window = [[NSWindow alloc] initWithContentRect:centerRect
+                                                   styleMask:NSTitledWindowMask | NSClosableWindowMask
+                                                     backing:NSBackingStoreBuffered
+                                                       defer:YES];
+    [window setDelegate:self];
+    
+    
+    
+    NSView* contentView = window.contentView;
+    [contentView setAutoresizesSubviews:YES];
+    
+    WebView* webView = [[WebView alloc] initWithFrame:contentView.frame];
+    [webView setFrameLoadDelegate:self];
+    [webView setResourceLoadDelegate:self];
+    [webView setPolicyDelegate:self];
+    [webView setAutoresizingMask:NSViewHeightSizable | NSViewWidthSizable];
+    
+    [contentView addSubview:webView];
+    
+    NSProgressIndicator* progressIndicator = [[NSProgressIndicator alloc] initWithFrame:NSMakeRect(DEFAULT_WINDOW_WIDTH / 2 - 16, DEFAULT_WINDOW_HEIGHT / 2 - 16, 32, 32)];
+    [progressIndicator setStyle:NSProgressIndicatorSpinningStyle];
+    // Keep the item centered in the window even if it's resized.
+    [progressIndicator setAutoresizingMask:NSViewMinXMargin | NSViewMaxXMargin | NSViewMinYMargin | NSViewMaxYMargin];
+    
+    // On OS X there's a noticable lag between the window showing and the page loading, so starting with the spinner
+    // at least make it looks liek something is happening.
+    [progressIndicator setHidden:NO];
+    [progressIndicator startAnimation:nil];
+    
+    [contentView addSubview:progressIndicator];
+    _progressIndicator = progressIndicator;
+    
+    _webView = webView;
+    self.window = window;
+    
+    return YES;
+}
+
+#pragma mark - UIViewController Methods
+
+#pragma mark - Event Handlers
+
+// Authentication was cancelled by the user by closing the window
+- (void)windowWillClose:(NSNotification *)notification
+{
+    (void)notification;
+    
+    [_delegate webAuthDidCancel];
+}
+
+// Fired 2 seconds after a page loads starts to show waiting indicator
+
+- (void)stop:(void (^)(void))completion
+{
+    [_webView.mainFrame stopLoading];
+    _delegate = nil;
+    [self close];
+    completion();
+}
+
+- (void)startRequest:(NSURLRequest *)request
+{
+    [self showWindow:nil];
+    [self loadRequest:request];
+}
+
+- (void)loadRequest:(NSURLRequest*)request
+{
+    [_webView.mainFrame loadRequest:request];
+}
+
+- (void)startSpinner
+{
+    [_progressIndicator setHidden:NO];
+    [_progressIndicator startAnimation:nil];
+    [self.window.contentView setNeedsDisplay:YES];
+}
+
+- (void)webView:(WebView *)webView decidePolicyForNavigationAction:(NSDictionary *)actionInformation
```

Comment:
Unfortunately in this case because "decidePolicyForNavigationAction:" is so much longer then "webView:" the preference is to keep it on one line as it's impossible to line up the colons, and the name of this method is coming from Apple's delegate protocol.